test/Dialect/FIRRTL/canonicalization.mlir

```diff
@@ -314,6 +316,10 @@ firrtl.module @Mux(in %in: !firrtl.uint<4>,
   %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
   %3 = firrtl.mux (%cond, %c1_ui1, %c1_ui0) : (!firrtl.uint<1>, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
   firrtl.connect %out1, %3 : !firrtl.uint<1>, !firrtl.uint<1>
+
+  // CHECK: firrtl.connect %out2, %c7_ui4
+  %4 = firrtl.mux (%condZeroWidth, %in, %c7_ui4) : (!firrtl.uint<0>, !firrtl.uint<4>, !firrtl.uint<4>) -> !firrtl.uint<4>
+  firrtl.connect %out2, %4 : !firrtl.uint<4>, !firrtl.uint<4>
```

Comment:
This patch is also allowing uninferred width condition right?  Please add a test for that.  Does inferwidths handle it correctly?  What if inferwidth infers it to 2 bits, do we produce an error?

Reply:
I can check how that works. Uninferred was allowed before. This is changing the type of the select argument from `UInt1Type` to `UInt01Type`. `UInt1Type` allowed for: (1) uninferred or (2) 1-bit. `UInt01Type` expands this to: (1) uninferred, (2) 0-bit, or (3) 1-bit. 

If it infers to 2-bit, that should be caught by the verifier due to `UInt1Type` or `UInt01Type` (I think). I'll check this.

Reply:
Cool thanks.  The problem with it being caught in the verifier is that that will crash the compiler, instead of reporting a new error back to the user.

Reply:
`InferWidths` treats the mux select line as a constraint that the width must be greater than or equal to one. If this infers to a width greater than one, it's will trip the verifier _inside `InferWidths`_ and produce (I thought) a fine warning? See below:

Example of current behavior:

```mlir
firrtl.circuit "Foo"   {
  firrtl.module @Foo(in %a: !firrtl.uint<2>, out %out: !firrtl.uint<1>) {
    %sel = firrtl.wire  : !firrtl.uint
    firrtl.connect %sel, %a : !firrtl.uint, !firrtl.uint<2>
    %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
    %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
    %0 = firrtl.mux(%sel, %c1_ui1, %c0_ui1) : (!firrtl.uint, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
    firrtl.connect %out, %0 : !firrtl.uint<1>, !firrtl.uint<1>
  }
}
```

```
# circt-opt -firrtl-infer-widths Foo.mlir
Foo.mlir:8:12: error: 'firrtl.mux' op operand #0 must be UInt<0> UInt<1> or UInt, but got '!firrtl.uint<2>'
      %0 = firrtl.mux(%sel, %c1_ui1, %c0_ui1) : (!firrtl.uint, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
           ^
Foo.mlir:8:12: note: see current operation: %0 = "firrtl.mux"(%sel, %c1_ui1, %c0_ui1) : (!firrtl.uint<2>, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
```

Can you clarify on what you'd like to see here?

There is already a check that this parses from `*.fir` and is allowed that was introduced in fb94acd3a9e886b08ea54494560a7644c3601082.

Reply:
I updated this PR to include the same check that a 0-bit select line parses.

Reply:
> Can you clarify on what you'd like to see here?

Exactly this, thank you!